### DISCUSSION TITLE
llamaindex: function_call does not generate subspans for a custom AsyncBaseTool class

### DISCUSSION THREAD

**Discussion by ootkin** [langfuse/langfuse#2298] [Support]
I'm using OpenAIAgent with a custom tool:
```py
from langfuse.llama_index import LlamaIndexCallbackHandler
from llama_index.core import Settings
from llama_index.core.callbacks import CallbackManager

langfuse_callback_handler = LlamaIndexCallbackHandler()
Settings.callback_manager = CallbackManager([langfuse_callback_handler])


class QueryPipelineTool(AsyncBaseTool):
    pipeline: QueryPipeline

    def call(self, input: Any) -> ToolOutput:
        pipeline_result = self.pipeline.run(query_str=input)
        return ToolOutput(
            content=str(pipeline_result),
            tool_name="query_pipeline_tool",
            raw_input={"query_str": input},
            raw_output=pipeline_result
        )

    async def acall(self, input: Any) -> ToolOutput:
        pipeline_result = await self.pipeline.arun(query_str=input)
        return ToolOutput(
            content=str(pipeline_result),
            tool_name="query_pipeline_tool",
            raw_input={"query_str": input},
            raw_output=pipeline_result
        )

    @property
    def metadata(self) -> ToolMetadata:
        return ToolMetadata(
            name="query_pipeline_tool",
            description="tool description",
        )

    @metadata.setter
    def metadata(self, value: ToolMetadata):
        self._custom_metadata = value

    def __init__(self, pipeline: QueryPipeline):
        self.pipeline = pipeline
        self._custom_metadata = None

tool = QueryPipelineTool(get_pipeline()) # this is a normal QueryPipeline from llamaindex
agent = OpenAIAgent.from_tools(
    tools=[tool],
    llm=OpenAI(model='gpt-4o', temperature=0),
)

agent.stream_chat(query)
```

And this is my trace:
![image](https://github.com/langfuse/langfuse/assets/22097635/4758ce3e-95d1-4a2f-85d9-a6d891b10199)

As you can see, the sub spans (for example retrievers steps) aren't generated in langfuse.

Could you please help me to solve this?
Thanks

**Comment by hassiebp** [comment]
Hi @ootkin - it seems like your custom `QueryPipelineTool` is not supporting / handling the callbacks that Langfuse relies on to provide detailed spans. Usually, in the llama-index project these context managers are used to dispatch events, [see here](https://github.com/run-llama/llama_index/blob/718ef2410ebc5947d983cce9477264bcabaae8e2/llama-index-core/llama_index/core/query_engine/custom.py#L39). Please open a question / issue in the llama-index project if you'd like to see your use case supported. 

The callback implementation is currently frozen and the llama-index project is moving towards the more flexible [instrumentation module](https://docs.llamaindex.ai/en/stable/module_guides/observability/instrumentation/). Once Langfuse has integrated with that, there will be more flexibility for you as well to extend your custom class' tracing.

**Reply by ootkin** [reply]
Thanks, i'll do it immediately.

**Comment by ootkin** [comment]
Hi, I just opened an issue on llama repo: https://github.com/run-llama/llama_index/issues/14097

Seems like I have to manually add the instrumentation inside the call method. How can I do it manually using llama_index?

**Comment by ootkin** [comment]
@hassiebp Could you please reopen this?

**Reply by marcklingen** [reply]
done
